test/e2e/upgrade_test.go

```diff
@@ -31,15 +31,69 @@ const (
 	windowsWorkloadTesterJob = "windows-workload-tester"
 )
 
+// getNodeTypeAffinities generate node affinity for non-BYOH and BYOH nodes if present
+func getNodeTypeAffinities() ([]*v1.Affinity, error) {
+	var affinities []*v1.Affinity
+	if gc.numberOfMachineNodes >= 1 {
+		affinityNoByoh, err := getNodeAffinityForLabel(v1.NodeSelectorOpDoesNotExist, controllers.BYOHLabel)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating node affinity without BYOH label")
+		}
+		affinities = append(affinities, affinityNoByoh)
+	}
+	if gc.numberOfBYOHNodes >= 1 {
+		affinityByoh, err := getNodeAffinityForLabel(v1.NodeSelectorOpIn, controllers.BYOHLabel, "true")
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating node affinity for BYOH label")
+		}
+		affinities = append(affinities, affinityByoh)
+	}
+	return affinities, nil
+}
+
+// generateLinuxWorkloadTesterCommand creates a Job to curl Windows webserver every 5 seconds.
+// If the webserver is not accessible for few minutes Curl Windows webserver
+// exit the pod with an error.
+func generateLinuxWorkloadTesterCommand(clusterIP string) []string {
+	return []string{
+		"bash",
+		"-c",
+		" while true;" +
+			" do curl " + clusterIP + ";" +
+			" if [ $? != 0 ]; then" +
+			" sleep 60;" +
+			" curl " + clusterIP + " || exit 1;" +
+			" fi; " +
+			"sleep 5; " +
+			"done"}
+}
+
 // upgradeTestSuite tests behaviour of the operator when an upgrade takes place.
 func upgradeTestSuite(t *testing.T) {
 	testCtx, err := NewTestContext()
 	require.NoError(t, err)
 
-	// test if Windows workloads are running by creating a Job that curls the workloads continuously.
-	testerJob, err := testCtx.deployWindowsWorkloadAndTester()
-	require.NoError(t, err, "error testing Windows workloads")
-	defer testCtx.deleteJob(testerJob.Name)
+	// generate node affinities
+	affinities, err := getNodeTypeAffinities()
+	require.NoError(t, err, "error getting node affinity")
+	// loop affinities to ensure deployed workloads will exclusively run based
+	// on the given affinity, e.g.  BYOH and/or Machine nodes
+	for _, affinity := range affinities {
+		// test if Windows workloads are running by creating a Job that curls the workloads continuously.
+		deployment, err := testCtx.deployWindowsWebServer("win-webserver", affinity)
+		require.NoErrorf(t, err, "error creating Windows Webserver deployment for upgrade test")
+		defer testCtx.deleteDeployment(deployment.Name)
+
+		// create a clusterIP service which can be used to reach the Windows webserver
+		intermediarySVC, err := testCtx.createService(deployment.Name, v1.ServiceTypeClusterIP, *deployment.Spec.Selector)
```

Comment:
Not sure why this intermediate service is needed.

Reply:
Same here

Reply:
Then why did you add it without understanding why this service is needed?

Reply:
It was added in https://github.com/openshift/windows-machine-config-operator/pull/148/commits/9d8001b47784c256bee666c72f6715cbdc2882e8

Reply:
That was added for a different reason. You seemed have copied it over without figuring if it is needed in your scenario or not.

Reply:
AFAIK, it is the same scenario.

Reply:
You are just wanting to ensure that pods get evicted. Do you need an intermediate service to ping?

Reply:
> You are just wanting to ensure that pods get evicted. Do you need an intermediate service to ping?

Correct, your statement is accurate. That was my initial implementation, but refactored a bit to reuse the [logic already present](https://github.com/openshift/windows-machine-config-operator/pull/645#discussion_r700390441)

Reply:
I can even understand if all the code was made common. Here it seems you reproduced what was being done in the windows exporter test case.

Reply:
Moreover, you could have potentially left the original workloads created in other tests running without having to create new ones here. And yes, there would be an ordering issue here. Not sure if that was not discussed before or not.

Reply:
@aravindhp Thanks for highlighting this.  No, the ordering issue was not discussed during the review. I feel relying on previously created workloads by other tests, introduces a hard dependency for the `e2e-upgrade` test, and may justify the creation of new workloads here.

Reply:
>I feel relying on previously created workloads by other tests, introduces a hard dependency for the e2e-upgrade test, and may justify the creation of new workloads here.

OK, that's a fair point and that's the statement I wanted to hear.